Enh/merge sequences

nathlacroix opened the pull request:
This pull-request addresses #68. It introduces new features to compress 2D sweeps while making most efficient use of the first dimension (Hardware sweep, i.e. sequence uploaded at once on all instruments). 

For instance, it should reduce a great part of the communication overhead when calibrating 2 qb gates and therefore allow for faster 2 qb gates tuneup. Can also be combined with parallel tuneup (not tested). 

Changes proposed in this pull request:
- Sequence.merge(): allows to merge sequences sequentially
- Deepcopy of sequences and segments: E.g. Now `deepcopy(segment)` works by avoiding the copy of the pulsar, which cannot be deepcopied
- Sequence.compress_2D_sweep(): compressed a 2D sweep easily with the list of sequences as input. See mqm.measure_cphase() for an example
- Adds the necessary parameter and reshaping in BaseAnalysis to decompress the data before proceeding to the usual analysis.

Current limitation:
- only works for sequences using the combination of SegmentHardSweep and SegmentSoftSweep for the 2D sweep (not instrument sweep functions).

Warning:
- features were tested separately but the entire measurement has not been tested on a physical setup yet.

Inviting @chellings for review. FYI @antsr @stephlazar  @christiankraglundandersen

nathlacroix committed:
Add merging feature for sequences

- Add merge function to Sequence, allows to merge several sequences into
  one or more, depending on the segment_limit
- Add deepcopy function to Segment and Sequence, as they cannot be
  blindly deepcopied because of the reference to the Pulsar instrument.

nathlacroix committed:
Enable easy compression of 2D sweeps

Add  a wrapper function of Sequence.merge to compress 2D sweeps as much
as possible to a lower number of sequences (merges subsequent sequences
together) while (i) keeping the total number of segments below a limit to
avoid memory overflow of UHF/HDAWG and (ii) ensuring all compressed
sequences have the same amount of segments (required to satisfy current
limitation of 2D sweep).

nathlacroix committed:
Add decompression of 2D sweeps to BaseAnalysis

- Gives the possibility to decompress 2D sweeps in which softsweep points
were converted into hardsweep point for the measurement back into the
shape before compression.
- return compression factor from sequence.compress_2D_sweep which has to
be passed as metadata to the analysis for decompression.

nathlacroix committed:
Add compression of 2D sweeps to CPhase function

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -293,20 +293,53 @@ def get_data_from_timestamp_list(self):
         if 'measured_data' in raw_data_dict and \
                 'value_names' in raw_data_dict:
             measured_data = raw_data_dict.pop('measured_data')
+            print(measured_data)
```
debug output

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -293,20 +293,53 @@ def get_data_from_timestamp_list(self):
             raw_data_dict['measured_data'] = OrderedDict()
 
             value_names = raw_data_dict['value_names']
+            print(value_names)
```
debug output

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -361,16 +392,21 @@ def extract_data(self):
+
             for i, rd_dict in enumerate(self.raw_data_dict):
                 temp_dict_list.append(
                     self.add_measured_data(rd_dict))
```
no compression factor here?

chellings commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -2255,7 +2255,11 @@ def measure_cphase(qbc, qbt, soft_sweep_params, cz_pulse_name,
+    if kw.get('compress_2D_sweep', False):
+        sequences, hard_sweep_points, soft_sweep_points, cf = \
+            sequences[0].compress_2D_sweep(sequences,
+                                           kw.get("compression_seg_lim", 300))
```
Do we really need 2 keyword args? Why not "if kw.get('compression_seg_lim', None) is not None"?

nathlacroix replied:
I think this would be compressing two distinct ideas into only one variable. What if we want to compress but without segment limit? Sure we could put it very high but it would be a bit hacky in my opinion

chellings commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -132,11 +140,186 @@ def repeat_ro(self, pulse_name, operation_dict):
                            (self.n_acq_elements(), 1))
+
+
+    @staticmethod
```
why are the 2 new functions staticmethods of the class? shouldn't they rather be just plain functions in the module?

nathlacroix replied:
Yes, good question, I also thought about that. They could be plain functions in the module. I just thought it would avoid loading the sequence module everytime we deal with sequences (e.g. in mqm where we get sequence objects back but typically do not construct sequence in, so the module is not loaded). With static functions, we can just access them from the returned sequence objects. 
Is this a sufficent use-case to have them part of the object? @chellings Otherwise, I'll move them

chellings replied:
OK, then keep them as static

chellings commented on pycqed/measurement/waveform_control/sequence.py:
```diff
@@ -132,11 +140,186 @@ def repeat_ro(self, pulse_name, operation_dict):
+                                              merge_repeat_patterns)
+        hard_sp_ind = np.arange(compressed_sequences[0].n_acq_elements())
+        soft_sp_ind = np.arange(len(compressed_sequences))
+        return compressed_sequences, hard_sp_ind, soft_sp_ind, factor
```
maybe compressed_sequences is a bit misleading. We do not really compress sequences, do we?

chellings reviewed (approved):
Thank you for implementing this. I am curious how much time it will save in an actual measurement. Apart from some minor comments I added, it looks good.

nathlacroix committed:
Address comments for pull request #69

- remove print statements
- add compression factor for multi-timestamp analysis
- fix get_param_value for multi-timestamp analysis

nathlacroix committed:
Address comments to pull request #69 (2)

- Add doc string for compression in measurement
- Merge two kwargs into only one to activate and specify the limit of
  the compression at the same time
